File: ObjC/notify-demo-app-objc/NotifyViewController.m
//
//  NotifyViewController.m
//  notify-demo-app-objc
//
//  Copyright © 2023 Mail.Ru LLC. All rights reserved.
//

#import "NotifyViewController.h"
#import "AppDelegate.h"
@import CXHubNotify;

static NSString *const DEFAULT_TEST_USER_ID = @"vladimir_2_test.gk.2011_dont_use_please@mail.ru";

@interface NotifyViewController ()

@end

@implementation NotifyViewController

- (IBAction)onChangeNotificationState:(id)sender {
    __auto_type alert = [UIAlertController alertControllerWithTitle:@"Notification state" message:nil preferredStyle:UIAlertControllerStyleActionSheet];

    [alert addAction:[UIAlertAction actionWithTitle:@"Enabled" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        CXNotify.getInstance.notificationState = CXNotificationStateEnabled;
    }]];
    [alert addAction:[UIAlertAction actionWithTitle:@"Only transactions" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        CXNotify.getInstance.notificationState = CXNotificationStateRestricted;
    }]];
    [alert addAction:[UIAlertAction actionWithTitle:@"Disabled" style:UIAlertActionStyleDestructive handler:^(UIAlertAction * _Nonnull action) {
        CXNotify.getInstance.notificationState = CXNotificationStateDisabled;
    }]];
    [self presentViewController:alert animated:true completion:nil];
}

- (IBAction)deviceIdTrackingSettings:(id)sender {
    __auto_type alert = [UIAlertController alertControllerWithTitle:@"DeviceID tracking" message:nil preferredStyle:UIAlertControllerStyleActionSheet];

    __auto_type actionEnable = [UIAlertAction actionWithTitle:@"Enable" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        [[CXNotify getInstance] setDeviceIdTrackingEnabled:true];
    }];
    [alert addAction:actionEnable];

    __auto_type actionDisable = [UIAlertAction actionWithTitle:@"Disable" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        [[CXNotify getInstance] setDeviceIdTrackingEnabled:false];
    }];
    [alert addAction:actionDisable];

    __auto_type actionCancel = [UIAlertAction actionWithTitle:@"Cancel" style:UIAlertActionStyleCancel handler:nil];
    [alert addAction:actionCancel];

    [self presentViewController:alert animated:true completion:nil];
}

- (IBAction)onSendTestEventsNowButtonClick:(id)sender {
    //An application could force CXHubSDK to send important events as soon as possible
    [[CXNotify getInstance] collectEventMap:@{@"TestEvent1":@"TestEventValue1", @"TestNumberEvent1":@(42)}
                          withImmediateLogic:YES];
    [[CXNotify getInstance] collectEvent:@"TestEvent2"
                                withValue:@"TestEventValue2"
                       withImmediateLogic:YES];
}

- (void)viewDidLoad {
    [super viewDidLoad];
    AppDelegate *appDelegate = (AppDelegate *)[[UIApplication sharedApplication] delegate];
    self.devicePushToken.text = appDelegate.devicePushToken;
    [[NSNotificationCenter defaultCenter] addObserverForName:TEST_APP_PUSH_TOKEN_OBTAINED object:nil queue:nil usingBlock:^(NSNotification * _Nonnull note) {
        self.devicePushToken.text = (NSString*) note.object;
    }];
    //Set user id after some delay to imitate user login process
    const dispatch_time_t popTime = dispatch_time(DISPATCH_TIME_NOW, 2);
    dispatch_after(popTime, dispatch_get_main_queue(), ^{
        //Inform CXHub about userId change
        [[CXNotify getInstance] setUserId:DEFAULT_TEST_USER_ID ofType:@"Email"];
    });
    self.instanceId.text = [[CXNotify getInstance] getInstanceId];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

@end

File: Swift/notify-demo-app-swift/NotifyViewController.swift
//
//  NotifyViewController.swift
//  notify-demo-app-swift
//
//  Copyright © 2023 Mail.Ru LLC. All rights reserved.
//

import UIKit
import CXHubNotify

class NotifyViewController: UIViewController, UITextFieldDelegate {
    
    private enum Constants {
        static let defaultTestUserId: String = "some@mail.ru"
        
        static let testEvent1: String = "TestEvent1"
        static let testEventValu1: NSString = NSString(string: "TestEventValue1")
        static let testNumberEvent1: String = "TestNumberEvent1"
        static let testNumberValue1: NSNumber = NSNumber(value: 42)
        
        static let testEvent2: String = "TestEvent2"
        static let testEventValu2: String = "TestEventValu2"
    }
    
    @IBOutlet weak var devicePushToken: UITextView!
    @IBOutlet weak var instanceId: UITextView!
    @IBOutlet weak var useSandbox: UILabel!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        guard let appDelegate = UIApplication.shared.delegate as? AppDelegate else {
            NSLog("Invalid app delegate")
            return
        }
        
        devicePushToken.text = appDelegate.devicePushToken
        NotificationCenter.default.addObserver(forName: .PushTokenObtained, object: nil, queue: nil) { (notification) in
            self.devicePushToken.text = (notification.object as? String) ?? nil
        }
        //Set user id after some delay to imitate user login process
        //DispatchQueue.main.asyncAfter(deadline: DispatchTime.now() + .seconds(5)) {
            //Inform CXHub about userId change
        //    CXNotify.getInstance()?.setUserId(Constants.defaultTestUserId)
        //}
        self.instanceId.text = CXNotify.getInstance()?.getInstanceId()
        var isUseSandboxString: String! = "YES"
        var isUseSandbox: Bool! = false
        isUseSandbox = CXApp.isAPNSSandbox
        if(!isUseSandbox) {
            isUseSandboxString = "NO"
        }
        self.useSandbox.text = isUseSandboxString
        let text = self.useSandbox.text
        NSLog("useSandbox: %@", text ?? "unknown")
        NSLog("", "")
    }
    

    @IBAction func changeNotificationState(sender: Any?) {
        let alert = UIAlertController(title: "Notification state", message: nil, preferredStyle: .actionSheet)
        alert.addAction(.init(title: "Enabled", style: .default, handler: { (_) in
            CXNotify.getInstance()?.notificationState = .enabled
        }))
        alert.addAction(.init(title: "Restricted", style: .default, handler: { (_) in
            CXNotify.getInstance()?.notificationState = .restricted
        }))
        alert.addAction(.init(title: "Disabled", style: .destructive, handler: { (_) in
            CXNotify.getInstance()?.notificationState = .disabled
        }))
        present(alert, animated: true, completion: nil)
    }

    @IBAction func deviceIDTrackingSettings(sender: Any?) {
        let alert = UIAlertController(title: "DeviceID tracking", message: nil, preferredStyle: .actionSheet)
        alert.addAction(.init(title: "Enable", style: .default, handler: { (_) in
            CXNotify.getInstance()?.setDeviceIdTrackingEnabled(true)
        }))
        alert.addAction(.init(title: "Disable", style: .default, handler: { (_) in
            CXNotify.getInstance()?.setDeviceIdTrackingEnabled(false)
        }))
        alert.addAction(.init(title: "Cancel", style: .cancel, handler: nil))
        present(alert, animated: true, completion: nil)
    }
    
    @IBAction func sendTestEventNow(sender: Any?) {
        //An application could force CXHubSDK to send important events as soon as possible
        CXNotify.getInstance()?.collectEventMap([Constants.testEvent1: Constants.testEventValu1, Constants.testNumberEvent1: Constants.testNumberValue1], withImmediateLogic: true)
        CXNotify.getInstance()?.collectEvent(Constants.testEvent2, withValue: Constants.testEventValu2, withImmediateLogic: true)
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
}
